### PR TITLE
Implement CMake build-chain for ASE + organize samples directory

### PR DESCRIPTION
**Pull request by abelardojarab**
WHAT THIS PULL REQUEST DOES?
=====================

The following set of commits enable:
* Build 4 ASE servers (each one corresponding for `NLB`, `hello_world`, `hello_world_mpf`, and `hello_afu`). `hello_afu` mimics a DCP setup, all the other 3 ASE server mimic an MCP setup.
* Generates coverage reports for ASE client using ASE-friendly version of `gtapi`, named `gtase`
   NOTE: (important), I made change so 'gtase` is built in addition to `gtapi` not as an alternative, before the tests were set that if you build 'gtase' then `gtapi` is not built.
* Usage of system compiler and linker instead of GCC included with Modelsim
* Reorganize all samples under a same 'samples' directory under `OPAE_SDK_ROOT`
* Enables building ASE and OPAE C++ framework completely independently of OPAE-C. It is possible to disable building OPAE-C by adding `-DBUILD_TOOLS=off` and `-DBUILD_LIBOPAE_C=off`

NEW DEFINED FLAGS FOR CMAKE SETUP (all are optional)
===========================
`-DBUILD_ASE_SAMPLES=on/off` --> enable building the ASE server for each one of the three samples mentioned above | off 
`-DBUILD_ASE_TESTS=on/off`  --> enable building ASE gtest application (gtase) 
`-DCMAKE_BUILD_TYPE=Coverage` --> will build all ASE server and client applications using Coverage flags (includes the SystemVerilog code)

Also remember to add these extra flags if you need to build ASE without OPAE-C.
`-DBUILD_TOOLS=off` and -`DBUILD_LIBOPAE_C=off.` 

**Note:** please dont disable `BUILD_LIBOPAE_C` and `BUILD_TOOLS` if you attempt to enable `BUILD_ASE_TESTS` since ASE tests use some code from OPAE tools as of now.

PRE-REQUISITES (before running cmake)
==========

* A valid installation of Modelsim SE 10.3 or higher
* A valid installation of Quartus 16 or higher

Also add this lines to your .bashrc (these variables are samples, your paths may differ for you)

```sh
# Modelsim options
export MODELSIMROOT=/tools/mentor/modelsim_se_10.5a
export MTI_HOME=$MODELSIMROOT/modeltech
export PATH=$MODELSIMROOT/modeltech/bin:$PATH
export PATH=$PATH:$MODELSIMROOT/modeltech/linux_x86_64
export MTI_VCO_MODE=64
export COMP64=1

# Altera options
export QUARTUS_HOME=/tools/altera/17.1.1.273/quartus
export QUARTUS_ROOTDIR=$QUARTUS_HOME
export QUARTUS_64BIT=1
export QUARTUS_ROOTDIR_OVERRIDE=$QUARTUS_HOME
export PATH="${QUARTUS_HOME}/bin:${QUARTUS_HOME}/sopc_builder/bin:${PATH}"

# Altera Quartus license server
export LM_LICENSE_FILE="${LM_LICENSE_FILE}:<ADD YOUR LMLICENSE SERVER HERE>"

# Modelsim license server
export MGLS_LICENSE_FILE="$MGLS_LICENSE_FILE:<ADD YOUR MGLS LICENSE SERVER HERE>"
```

SAMPLE CMAKE SETUP (BUILDING ASE SAMPLES in addition to OPAE-C)
==========
Thus a sample cmake setup which builds all ASE server and client applications and ASE tests is as follows:

```console
$ cmake -DBUILD_ASE_SAMPLES=on \
           -DBUILD_ASE_TESTS=on \
           -DCMAKE_BUILD_TYPE=Coverage 
            <directory where opae-sdk is cloned>
```
sample hello_fpga written in c running on ase server for intg_nlb0 setup

SAMPLE CMAKE SETUP (BUILDING ASE SAMPLES without OPAE-C)
==========

```console
$ cmake -DBUILD_LIBOPAE_C=off -DBUILD_TOOLS=off -DBUILD_ASE_SAMPLES=on ~/workspace_remote/opae-sdk/
```

**Note:** Disabling building OPAE tools (`-DBUILD_TOOLS=off`) disables building ASE tests too.

HOW TO RUN THE ASE SERVERS?
======================
Under your build directory you will find a 'samples' directory and inside 'samples' you will find built files for each sample

```console
[12:03:25] abelardojara@ubuntu01 [~/cmake_builds/opae-sdk/samples] $ tree -d -L 1
.
├── base
│   ├── hw
│   └── sw
├── hello_afu
│   ├── hw
│   └── sw
├── hello_world
│   ├── hw
│   └── sw
├── hello_world_mpf
│   ├── hw
│   └── sw
└── intg_xeon_nlb
    └── hw
```

Inside each `hw` directory you will find an executable script name `ase_server.sh`. For each example, just run that file `ase_server.sh` to launch the corresponding ASE server.

HOW ABOUT THE SW ASE CLIENTS?
=====================

They are all located under 'bin' directory inside your build directory.

The files are:
```
hello_fpga --> uses intg_xeon_nlb ASE server
hello_world --> uses hello_world ASE server
hello_world_mpf  --> uses hello_world_mpf ASE server
hello_afu ---> uses hello_afu ASE server
gtase --> this is the ASE gtest application which uses intg_xeon_nlb example ASE server
```

### CODE REVIEW

**Comment by lzhan55** [comment]
Did you run the DCP tests?

**Comment by abelardojarab** [comment]
@lzhan55 hi Tracy
no it does not install anything. I havent added any installation target for the .deb or .rpm.

**Comment by lzhan55** [comment]
I checked out the branch and tried to build the ase client using "cmake ../ -DBUILD_ASE=YES -DCMAKE_INSTALL_PREFIX=/tmp/opae/myinst/", I got error in the attached image cmake_opae_orig.PNG

I got the error in the attached file cmake_ase_server.PNG with your new build command "cmake -DBUILD_ASE_SAMPLES=on -DBUILD_TOOLS=on -DCMAKE_BUILD_TYPE=Coverage -DBUILD_ASE_TESTS=on .. "

cmake_opae_orig
cmake_ase_server

**Comment by abelardojarab** [comment]
@lzhan55 good catch .. seems that i set file permissions wrong. i will fix first thing on Monday.
thanks a lot. A.

**Comment by lzhan55** [comment]
I checked out cmake_ase_server branch and built and ran all the samples successfully.

**Comment by zhangzhang10** [comment]
Do you also update the documentation to reflect this new changes?

**Comment by lzhan55** [comment]
@abelardojarab Before you check in the code to trunk, you should run DCP ASE regression tests. Thanks.

**Comment by nakulkorde** [comment]
This should go through our regression tests as well. Once I have the build ready, I am still working on the build config for that, once done, I will run the tests.

**Comment by lzhan55** [comment]
DCP regression tests http://sw-pert.altera.com/pert/pert.php?test_run_id=2380934

File: samples/hello_afu/hw/rtl/afu_json_info.vh [hunk 1]
```diff
@@ -0,0 +1,13 @@
+//
+// Generated by afu_json_mgr from cci_hello.json
+//
+
+`ifndef __AFU_JSON_INFO__
+`define __AFU_JSON_INFO__
+
+`define AFU_ACCEL_NAME "hello_afu"
+`define AFU_ACCEL_UUID 128'hc6aa954a_9b91_4a37_abc1_1d9f0709dcc3
```
Comment (lzhan55) [inline samples/hello_afu/hw/rtl/afu_json_info.vh]: hello_afu UUID incorrect here, not the same one from json file. This one is hello_world's AFU ID and that is the reason why hello_afu test failed.

Comment (lzhan55) [reply samples/hello_afu/hw/rtl/afu_json_info.vh]: Hello AFU ID should be 128'h1cddcf01_c266_44de_af63_9d09858310e5

**Pull request closed without merging**